### ISSUE TITLE
Rule proposal: No return promise within try catch

### ISSUE THREAD

**Issue by danny-may** [typescript-eslint/typescript-eslint#4322]
### Description

If you have an async call placed inside a try/catch block and return the promise rather than the awaited result then any errors that are thrown asyncronously will not be caught by the try/catch. While this behaviour can be useful, it can be very easy to miss as returning promises from an async method is something js/ts allow and so dont warn you about it.

This would also be solved by disallowing returning promises from async methods altogether, but that might be a bit heavy handed.

### Fail

```ts
const myCache = new Map<string, MyRecord>();
async function getRecordAPI(is: string): Promise<MyRecord> {
  // Do something that may throw async
  throw new Error('Record not found');
}

async function tryGetRecord(id: string): Promise<MyRecord | undefined> {
  try {
    return myCache.get(id) ?? getRecordAPI(id); // <--- issue is no await here
  } catch (error: unknown) {
    if (error instanceof Error && error.message === 'Record not found')
      return undefined; // This will never be hit
    throw error;
  }
}
```

### Pass

```ts
const myCache = new Map<string, MyRecord>();
async function getRecordAPI(is: string): Promise<MyRecord> {
  // Do something that may throw async
  throw new Error('Record not found');
}

async function tryGetRecord(id: string): Promise<MyRecord | undefined> {
  try {
    return myCache.get(id) ?? await getRecordAPI(id);
  } catch (error: unknown) {
    if (error instanceof Error && error.message === 'Record not found')
      return undefined;
    throw error;
  }
}
```

**Comment by bradzacher** [comment]
https://typescript-eslint.io/rules/return-await
